### ISSUE TITLE
Issues in v1.4.0 migration from 1.3

### ISSUE THREAD

**Issue by jlfernandez-cba** [opencb/opencga#1226]
The migration from opencga 1.3 to opencga 1.4 has some problems. The most important is the incorrect assignation of uid's from id's, as IDs are String in opencga 1.3, and UIDs are Int64 (migration.js). This produces an exception, which is visible when calling the REST services, but which is masked when using the CLI, and the user simply gets no results, with no apparent error. Correct integer UIDs should be generated. For example,

https://github.com/opencb/opencga/blob/5e422782991298fd6f2267c426f0c15527809292/opencga-app/app/migration/v1.4.0/catalog/migration.js#L502

Another, unrelated problem, is the wrong assumption that Job.stdOutput and stdError cannot be null. This null values results from most job generating commands, but

https://github.com/opencb/opencga/blob/5e422782991298fd6f2267c426f0c15527809292/opencga-app/app/migration/v1.4.0/catalog/migration.js#L440

fails in this case, which prevent Jobs to migrate at all.

Finally,

https://github.com/opencb/opencga/blob/5e422782991298fd6f2267c426f0c15527809292/opencga-app/app/migration/v1.4.0/catalog/migration.js#L244

uses unsetChanges without previously declaring it, throwing an exception when executing the script. It should be declared after setChanges, as in other functions.

**Comment by pfurio** [comment]
Thanks for reporting those issues.

Fortunately, the first issue that seemed the most worrying is not an issue at all. Id's in general, and the project id in particular, have always been treated as _longs_ in OpenCGA 1.3 (https://github.com/opencb/opencga/blob/hotfix-1.3.x/opencga-core/src/main/java/org/opencb/opencga/core/models/Project.java#L28) so that line you showed should always work assuming the db was actually using 1.3. Maybe you have been jumping between versions and there was an inconsistency?

I will definitely fixed the other two. Many thanks

**Comment by jlfernandez-cba** [comment]
A pleasure to help.

I checked the source of the problem you comment. While I was testing and finding all the problems I mentioned this morning (v1.4.0 not being able to index VCF files, not listing projects but showing no errors, the other bugs in migration, and so on), I ran several times the migration procedure, iteratively fixing these bugs. Of course, this assumes that if the migration procedure fails, it will leave the database in a state such that migration can be resumed (for example, checking whether the IDs have already being converted). Obviously this is not the case, which makes the debugger miserable. The IDs were strings because OpenCGA 1.4 migrated part of the database and then failed (for several reasons, as I showed), leaving the database in an inconsistent state for itself, and not because of a previous installation. It's now clear that, until we debug all the problems with the migration, I'll have to restore the complete database every time I fix a bug.

Unfortunately, and contrary to your assertion, this makes any other bugs in migration equally important, as any bug will require a complete restoration of the database. More careful testing is required before trying this on production.
